eviloauth/dispatcher.py

```diff
@@ -1,86 +1,297 @@
 import sys
 import logging
+from diskcache import Cache
 from eviloauth.idp import IDP
 from eviloauth.exceptions import EviloauthCommandException
+from eviloauth.access_token import AccessToken
+from eviloauth import app, cache, temp_dir, load_modules
+from datetime import datetime, timedelta
 
+class TokenManager:
+    def __init__(self, cache_dir, general_cache):
+        self.cache = Cache(cache_dir)
+        self.general_cache = general_cache
 
-class Dispatcher:
-    def __init__(self, flask_server, module_dict, cache, redirect_server):
-        logging.debug('Initializing dispatcher')
-        logging.debug(f'\tFlask server: {flask_server}')
-        logging.debug(f'\tModule dict: {module_dict}')
-        logging.debug(f'\tCache: {cache}')
-        logging.debug(f'\tRedirect server: {redirect_server}')
+    def list_tokens(self):
+        general_tokens = self.general_cache.get('tokens', {})
+        return list(general_tokens.keys())
+    
+    def add(self, access_token):
+        if isinstance(access_token, AccessToken):
+            token_key = str(access_token)
+            self.cache.set(token_key, access_token)
+            return token_key
+            
+    def delete(self, token_key):
+            if token_key in self.cache:
+                self.cache.delete(token_key)
+                return True
+            return False
 
+    def list(self):
+        return [key for key in self.cache.keys()]
+        
+class Dispatcher:
+    def __init__(self, flask_server, module_dict, redirect_server, cache):
         self.flask_server = flask_server
         self.module_dict = module_dict
-        self.cache = cache
         self.redirect_server = redirect_server
+        self.cache = cache
+        self.token_manager = TokenManager(temp_dir, cache) 
+        
+    @classmethod
+    def get_token_manager(cls):
+        if cls.instance:
+            return cls.instance.token_manager
+        else:
+            return None
 
     def dispatch(self, commands):
-        cmd, sub, arg, *args = commands.split(' ') + [None, None, None]
+        cmd_parts = commands.split(' ')
```

Comment:
You've added a bug by doing this. Change it back to the original.

If no commands are passed, the program crashes.